src/electron/future/services/download-manager/index.ts

```diff
@@ -0,0 +1,181 @@
+ * are centrally managed. The class provides methods to manipulate the download queue, including adding new downloads and querying the queue status.
+ * It automatically handles the processing of queued downloads, respecting the maximum concurrent downloads limit and updating the status of each download accordingly.
+ */
+export class DownloadManager {
+	private static instance: DownloadManager | null;
+	private downloadQueue: Array<DownloadItem>;
+	private currentDownloads = 0;
+	private maxConcurrentDownloads = 1;
+
+	/**
+	 * Constructs a new instance of the DownloadManager class.
+	 */
+	constructor() {
+		this.downloadQueue = [];
+	}
+
+	/**
+	 * Retrieves the singleton instance of the DownloadManager.
+	 * @returns The singleton instance of the DownloadManager.
+	 */
+	public static getInstance(): DownloadManager {
+		DownloadManager.instance ||= new DownloadManager();
+
+		return DownloadManager.instance;
+	}
+
+	/**
+	 * Resets the singleton instance of the DownloadManager to null.
+	 */
+	public static resetInstance(): void {
+		DownloadManager.instance = null;
+	}
+
+	/**
+	 * Gets the current size of the download queue.
+	 * @returns The number of items in the download queue.
+	 */
+	public getQueueSize() {
+		return this.downloadQueue.length;
+	}
+
+	/**
+	 * Retrieves the current download queue.
+	 * @returns An array of DownloadItem objects in the queue.
+	 */
+	public getDownloadQueue() {
+		return this.downloadQueue;
+	}
+
+	/**
+	 * Checks if an item is already present in the download queue.
+	 * @param itemId The unique identifier of the download item.
+	 * @returns A boolean indicating whether the item is in the queue.
+	 */
+	public isItemInQueue(itemId: string): boolean {
+		return [...this.downloadQueue].some(item => item.id === itemId);
+	}
+
+	/**
+	 * Adds a new item to the download queue and processes the queue if possible.
+	 * @param item The DownloadItem object to add to the queue.
+	 */
+	public addToQueue(item: DownloadItem): void {
+		item.state = DownloadState.WAITING;
+		const isDuplicate = this.downloadQueue.some(existingItem => existingItem.id === item.id);
+		if (!isDuplicate) {
+			this.downloadQueue.push(item);
+			this.processQueue();
+		}
+	}
+
+	/**
+	 * Processes the next item in the download queue if the maximum concurrent downloads limit has not been reached.
+	 * @private
+	 */
+	private async processQueue(): Promise<void> {
+		// If there's already the maximum number of downloads in progress, do nothing
+		if (this.currentDownloads >= this.maxConcurrentDownloads) {
+			return;
+		}
+
+		const nextItem = this.downloadQueue.find(item => item.state === DownloadState.WAITING);
+		if (!nextItem) {
+			console.log("No more items to process");
+			return;
+		}
+
+		await this.startDownload(nextItem);
+	}
+
+	/**
+	 * Starts downloading an item, updating its state throughout the download process, and handling completion, progress, and errors.
+	 * @param item The DownloadItem to be downloaded.
```

Comment:
It would be nice to know in the tsdocs, that when a download is completed, that the queue will be processed automatically again.

Reply:
I know that this becomes clear when looking at the code, it would be just nice for people to have this information more direct.